extension/ballastextension/README.md

```diff
@@ -53,6 +53,9 @@ Config that uses 64 Mib of memory for the ballast:
 extensions:
   memory_ballast:
     size_mib: 64
+
+service:
+  extensions: [memory_ballast]
```

Comment:
We don't generally have this for any other component, e.g. see https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/otlpexporter

I would suggest removing this:

```suggestion
```

And add some wording like:

> Note that you need to **add the extension to the `services::extensions` list** for it to be enabled

instead. Do you think that is sufficiently clear?

Reply:
Either approach sounds good, but this is helpful for lazy developers that just want to copy/paste some yaml for the extension to work. They might miss that extra note and then waste time troubleshooting like I did. It's not intuitive that configuring an extension doesn't enable it

Reply:
> Either approach sounds good, but this is helpful for lazy developers that just want to copy/paste some yaml for the extension to work

I understand, it just feels weird to do this only for extensions. All components in the OpenTelemetry Collector have this property, and we are not documenting this for any other kind of component. Why are extensions special in this sense?

>  They might miss that extra note and then waste time troubleshooting like I did.

I feel like the way to do this is with a warning, strict mode or some other kind of automated validation

> It's not intuitive that configuring an extension doesn't enable it

Changing the default behavior on this would be a big breaking change. If you have suggestions on a backwards-compatible way of doing this, a migration plan or a tool to help with this, I am happy to read a proposal, but in general this is hard to change today.

Reply:
The other triager in https://github.com/open-telemetry/opentelemetry-collector/issues/9195 also didn't think my approach was the right one, so all good to close this suggestion

> I understand, it just feels weird to do this only for extensions. All components in the OpenTelemetry Collector have this property, and we are not documenting this for any other kind of component. Why are extensions special in this sense?

I think it's more intuitive for other components since you need to pick a pipeline, and pipelines are a more well-understood concept than extensions. 

Extensions were originally listed under services to control the order they are started and stopped, according to the README. However, I don't think the order matters for most if not all the extensions, so people are less likely to think about that compared to pipelines

>I feel like the way to do this is with a warning, strict mode or some other kind of automated validation
I think the lazy developer would also miss the documentation for strict mode :)

>Changing the default behavior on this would be a big breaking change. If you have suggestions on a backwards-compatible way of doing this, a migration plan or a tool to help with this, I am happy to read a proposal, but in general this is hard to change today.

Yeah this would probably be overkill to change, but one idea would be to fallback to the order that extensions are configured if you forget to set the order under `services`

Reply:
> The other triager in https://github.com/open-telemetry/opentelemetry-collector/issues/9195 also didn't think my approach was the right one, so all good to close this suggestion

We can wait to see what others think, but in its current state I think it would leave our documentation in an inconsistent state and I would like to avoid that.

> Extensions were originally listed under services to control the order they are started and stopped, according to the README. However, I don't think the order matters for most if not all the extensions, so people are less likely to think about that compared to pipelines
> 
> > I feel like the way to do this is with a warning, strict mode or some other kind of automated validation
> 
> I think the lazy developer would also miss the documentation for strict mode :). A warning log on startup would be helpful

:+1: I think a warning is likely to be less controversial than this.

> fallback to the order that extensions are configured if you forget to set the order under services

YAML keys [don't have a defined order](https://yaml.org/spec/1.2.2/#3221-mapping-key-order), so I don't think this is a good idea (for example, AIUI intermediate steps from the YAML the user writes until it gets to the Collector are free to change the order of the keys per the YAML spec). Also in practical terms, any Go YAML parser out there (at least that I know of) will map this to a Go map, which also does not preserve the order of the keys, so we would need to write some custom code and custom structure to save the map (which is both a lot of work and would need us to rethink the API we offer users).